File: leetcode-go/binary_search/median_of_two_sorted_arrays.go
// Package binary_search https://leetcode-cn.com/problems/median-of-two-sorted-arrays/
package binary_search

func min(n1, n2 int) int {
	if n1 < n2 {
		return n1
	} else {
		return n2
	}
}

func getKth(nums1 []int, nums2 []int, k int) int {
	len1 := len(nums1)
	len2 := len(nums2)

	index1 := 0
	index2 := 0

	for {
		if index1 >= len1 {
			return nums2[index2+k-1]
		}
		if index2 >= len2 {
			return nums1[index1+k-1]
		}
		if k == 1 {
			return min(nums1[index1], nums2[index2])
		}

		halfK := k / 2

		half1 := min(index1+halfK, len1) - 1
		half2 := min(index2+halfK, len2) - 1

		pivot1 := nums1[half1]
		pivot2 := nums2[half2]

		if pivot1 < pivot2 {
			k -= half1 - index1 + 1
			index1 = half1 + 1
		} else {
			k -= half2 - index2 + 1
			index2 = half2 + 1
		}
	}

}

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
	len1 := len(nums1)
	len2 := len(nums2)
	total := len1 + len2
	if total%2 == 1 {
		return float64(getKth(nums1, nums2, total/2+1))
	} else {
		return float64(getKth(nums1, nums2, total/2)+getKth(nums1, nums2, total/2+1)) / 2.0
	}
}

File: leetcode-rust/solutions/src/binary_search/median_of_two_sorted_arrays.rs
// https://leetcode.com/problems/median-of-two-sorted-arrays/description/
struct Solution;

impl Solution {
    pub fn find_median_sorted_arrays(nums1: Vec<i32>, nums2: Vec<i32>) -> f64 {
        let (n1, n2) = (nums1.len(), nums2.len());
        if n1 > n2 {
            return Solution::find_median_sorted_arrays(nums2, nums1);
        }
        let mut left = 0;
        let mut right = n1;
        while left <= right {
            let mid1 = (left + right) / 2;
            let mid2 = (n1 + n2 + 1) / 2 - mid1;
            let left1 = if mid1 == 0 { i32::MIN } else { nums1[mid1 - 1] };
            let right1 = if mid1 == n1 { i32::MAX } else { nums1[mid1] };
            let left2 = if mid2 == 0 { i32::MIN } else { nums2[mid2 - 1] };
            let right2 = if mid2 == n2 { i32::MAX } else { nums2[mid2] };
            if left1 <= right2 && left2 <= right1 {
                if (n1 + n2) % 2 == 0 {
                    return (left1.max(left2) as f64 + right1.min(right2) as f64) / 2.0;
                } else {
                    return left1.max(left2) as f64;
                }
            } else if left1 > right2 {
                right = mid1 - 1;
            } else {
                left = mid1 + 1;
            }
        }
        unreachable!();
    }
}

#[cfg(test)]
mod tests {
    use crate::binary_search::median_of_two_sorted_arrays::Solution;

    #[test]
    fn case01() {
        let nums1 = vec![1, 2];
        let nums2 = vec![3, 4];
        let actual = Solution::find_median_sorted_arrays(nums1, nums2);
        let expected = 2.5;
        assert_eq!(actual, expected);
    }
}